src/weightedcoorddesc.jl

```diff
@@ -0,0 +1,88 @@
+mutable struct WeightedCoordinateDescent{T}
+    M::Matrix{T}
+    maxiter::Int           # maximum number of iterations (in main procedure)
+    verbose::Bool          # whether to show procedural information
+    tol::T                 # tolerance of changes on W and H triggering convergence
+    update_H::Bool         # whether to update H
+    shuffle::Bool          # # if true, randomize the order of coordinates in the CD solver           # mask matrix
+    function WeightedCoordinateDescent{T}(M::AbstractMatrix;
+                                        maxiter::Integer=100,
+                                        verbose::Bool=false,
+                                        tol::Real=cbrt(eps(T)),
+                                        update_H::Bool=true,
+                                        shuffle::Bool=false) where T
+        new{T}(M, maxiter, verbose, tol, update_H, shuffle)
+    end
+end
+
+
+solve!(alg::WeightedCoordinateDescent{T}, X, W, H) where {T} =
+    nmf_skeleton!(WeightedCoordinateDescentUpd{T}(alg.M, alg.shuffle, alg.update_H),
+                  X, W, H, alg.maxiter, alg.verbose, alg.tol)
+
+
+struct WeightedCoordinateDescentUpd{T} <: NMFUpdater{T}
+    M::Matrix{T}
+    shuffle::Bool
+    update_H::Bool
+    function WeightedCoordinateDescentUpd{T}(M::Matrix{T}, shuffle::Bool, update_H::Bool) where {T}
+        new{T}(M,
+               shuffle,
+               update_H)
+    end
+end
+
+mutable struct WeightedCoordinateDescentState{T}
+    WH::Matrix{T}
+
+    function WeightedCoordinateDescentState{T}(W, H) where T
+        new{T}(W*H)
+    end
+end
+
+prepare_state(::WeightedCoordinateDescentUpd{T}, ::AbstractArray{T}, W, H) where T = WeightedCoordinateDescentState{T}(W, H)
+
+function evaluate_objv(u::WeightedCoordinateDescentUpd{T}, s::WeightedCoordinateDescentState{T}, X, W, H) where T
+    M = u.M
+    mul!(s.WH, W, H)
+    convert(T, 0.5) * sqL2dist(M.*X, M.*s.WH)
```

Comment:
```suggestion
    sqL2dist(M.*X, M.*s.WH) / 2
```

Then you don't need `T` at all.